In [3]:
import pandas as pd
import pandas_profiling
import numpy as np
import seaborn as sns
from pandas_profiling import ProfileReport
from sklearn.decomposition import PCA
import warnings
warnings.filterwarnings("ignore")

In [4]:
df = pd.read_csv('FS_2y_before_Bankruptcy_train_preprocessed.csv',encoding='utf-8')


In [5]:
df_liq = df[['Attr4','Attr40']]
df_op_prof = df[['Attr42','Attr49']]
df_bsns_rsk = df[['Attr13','Attr19']]
df_fncl_rsk = df[['Attr10','Attr2','Attr25']]


In [6]:
from sklearn.preprocessing import StandardScaler
df_liq = StandardScaler().fit_transform(df_liq)
df_op_prof = StandardScaler().fit_transform(df_op_prof)
df_bsns_rsk = StandardScaler().fit_transform(df_bsns_rsk)
df_fncl_rsk = StandardScaler().fit_transform(df_fncl_rsk)

In [7]:
pca = PCA(n_components=df_liq.shape[1])
pca.fit(df_liq)
explained_variance_ratio = pca.explained_variance_ratio_
pd.DataFrame(np.cumsum(explained_variance_ratio[:]),  
                    columns=['Cum_V_ration'])


Cum_V_ration
0       0.86701
1       1.00000

In [8]:
pcs = pca.fit_transform(df_liq)
n_of_pcs = 1
print("PC의 설명력 : {:.3f}%".format(np.sum(explained_variance_ratio[:n_of_pcs]*100)))
df_liq_pc = pd.DataFrame(pcs[:,:n_of_pcs], 
                        columns = ['liq_1'][:n_of_pcs])

PC의 설명력 : 86.701%


op_prof

In [9]:
pca = PCA(n_components=df_op_prof.shape[1])
pca.fit(df_op_prof)
explained_variance_ratio = pca.explained_variance_ratio_
pd.DataFrame(np.cumsum(explained_variance_ratio[:]),  
                    columns=['Cum_V_ration'])

Cum_V_ration
0      0.942934
1      1.000000

In [10]:
pcs = pca.fit_transform(df_op_prof)
n_of_pcs = 1
print("PC의 설명력 : {:.3f}%".format(np.sum(explained_variance_ratio[:n_of_pcs]*100)))
df_op_prof_pc = pd.DataFrame(pcs[:,:n_of_pcs], 
                        columns = ['op_1'][:n_of_pcs])

PC의 설명력 : 94.293%


bsns_rsk

In [11]:
pca = PCA(n_components=df_bsns_rsk.shape[1])
pca.fit(df_bsns_rsk)
explained_variance_ratio = pca.explained_variance_ratio_
pd.DataFrame(np.cumsum(explained_variance_ratio[:]), 
                    columns=['Cum_V_ration'])

Cum_V_ration
0      0.785348
1      1.000000

In [12]:
pcs = pca.fit_transform(df_bsns_rsk)
n_of_pcs = 1
print("PC의 설명력 : {:.3f}%".format(np.sum(explained_variance_ratio[:n_of_pcs]*100)))
df_bsns_rsk_pc = pd.DataFrame(pcs[:,:n_of_pcs], 
                        columns = ['bs_1'][:n_of_pcs])

PC의 설명력 : 78.535%


fncl_rsk

In [13]:
pca = PCA(n_components=df_fncl_rsk.shape[1])
pca.fit(df_fncl_rsk)
explained_variance_ratio = pca.explained_variance_ratio_
pd.DataFrame(np.cumsum(explained_variance_ratio[:]), 
                    columns=['Cum_V_ration'])

Cum_V_ration
0      0.938459
1      0.991801
2      1.000000

In [14]:
pcs = pca.fit_transform(df_fncl_rsk)
n_of_pcs = 1
print("PC의 설명력 : {:.3f}%".format(np.sum(explained_variance_ratio[:n_of_pcs]*100)))
df_fncl_rsk_pc = pd.DataFrame(pcs[:,:n_of_pcs], 
                        columns = ['fn_1'][:n_of_pcs])

PC의 설명력 : 93.846%


concatenate

In [15]:
df_list = [df_liq_pc, df_op_prof_pc,df_bsns_rsk_pc,df_fncl_rsk_pc,df ]
df_full = pd.concat(df_list, ignore_index=False,axis=1)
df_full.head()

# 35개

liq_1      op_1      bs_1      fn_1  Unnamed: 0     Attr1    Attr2  \
0 -5.690372e-02 -0.039632 -0.006127 -0.527226           0 -0.024558  0.30051   
1  4.056575e-17 -0.097482 -0.038243 -0.412242           1  0.639180  0.00000   
2 -5.763256e-02 -0.015167  0.006112 -0.194449           2 -0.440760  0.47723   
3 -2.880104e-03 -0.046792 -0.011557 -0.772702           3  0.119520  0.17679   
4 -3.597121e-02 -0.068734 -0.023606 -0.562140           4  0.230400  0.27095   

     Attr3     Attr4    Attr5  ...    Attr56    Attr57   Attr58   Attr59  \
0  0.10733  1.836800   34.382  ...  0.062497 -0.036289  0.93750  0.25454   
1  1.00000  8.664398   29.678  ...  0.679830  0.639180  0.32014  0.00000   
2  0.19260  1.467800   34.555  ... -0.290590 -0.843130  1.28710  0.12538   
3  0.63546  4.594400  117.650  ...  0.112110  0.147520  0.88789  0.00000   
4  0.23762  2.574500  -26.928  ...  0.459570  0.316020  0.62320  0.16465   

       Attr60   Attr61  Attr62   Attr63     Attr64  class  
0   13.632000   3.6930  69.389  5.26020   0.882600    0.0  
1  134.817648  37.8860   0.000  8.99491  36.363852    0.0  
2  134.817648   2.5649  98.950  3.68870   3.837600    0.0  
3    8.930200  10.2870  40.355  9.04480   8.516900    0.0  
4    7.927400   6.6605  68.502  5.32830   1.315000    0.0  

[5 rows x 69 columns]

In [16]:
df_full.drop(['Attr4','Attr40','Attr42','Attr49','Attr13','Attr19','Attr10','Attr2','Attr25'],axis=1,inplace=True)

df_full1 = df_full.sample(frac=0.95, random_state=786)
df_full1.reset_index(inplace=True, drop=True)

df_unseen = df_full.drop(df_full1.index)
df_unseen.reset_index(inplace=True, drop=True)

df_full.head()

liq_1      op_1      bs_1      fn_1  Unnamed: 0     Attr1    Attr3  \
0 -5.690372e-02 -0.039632 -0.006127 -0.527226           0 -0.024558  0.10733   
1  4.056575e-17 -0.097482 -0.038243 -0.412242           1  0.639180  1.00000   
2 -5.763256e-02 -0.015167  0.006112 -0.194449           2 -0.440760  0.19260   
3 -2.880104e-03 -0.046792 -0.011557 -0.772702           3  0.119520  0.63546   
4 -3.597121e-02 -0.068734 -0.023606 -0.562140           4  0.230400  0.23762   

     Attr5     Attr6     Attr7  ...    Attr56    Attr57   Attr58   Attr59  \
0   34.382 -0.026711 -0.020067  ...  0.062497 -0.036289  0.93750  0.25454   
1   29.678 -1.139300  0.760520  ...  0.679830  0.639180  0.32014  0.00000   
2   34.555  0.000000 -0.440760  ... -0.290590 -0.843130  1.28710  0.12538   
3  117.650  0.251540  0.148750  ...  0.112110  0.147520  0.88789  0.00000   
4  -26.928  0.617540  0.282690  ...  0.459570  0.316020  0.62320  0.16465   

       Attr60   Attr61  Attr62   Attr63     Attr64  class  
0   13.632000   3.6930  69.389  5.26020   0.882600    0.0  
1  134.817648  37.8860   0.000  8.99491  36.363852    0.0  
2  134.817648   2.5649  98.950  3.68870   3.837600    0.0  
3    8.930200  10.2870  40.355  9.04480   8.516900    0.0  
4    7.927400   6.6605  68.502  5.32830   1.315000    0.0  

[5 rows x 60 columns]

In [17]:
df_unseen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341 entries, 0 to 340
Data columns (total 60 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   liq_1       341 non-null    float64
 1   op_1        341 non-null    float64
 2   bs_1        341 non-null    float64
 3   fn_1        341 non-null    float64
 4   Unnamed: 0  341 non-null    int64  
 5   Attr1       341 non-null    float64
 6   Attr3       341 non-null    float64
 7   Attr5       341 non-null    float64
 8   Attr6       341 non-null    float64
 9   Attr7       341 non-null    float64
 10  Attr8       341 non-null    float64
 11  Attr9       341 non-null    float64
 12  Attr11      341 non-null    float64
 13  Attr12      341 non-null    float64
 14  Attr14      341 non-null    float64
 15  Attr15      341 non-null    float64
 16  Attr16      341 non-null    float64
 17  Attr17      341 non-null    float64
 18  Attr18      341 non-null    float64
 19  Attr20      341 non-null    f

In [18]:
from pycaret.classification import *

exp_name = setup(data = df_full1,  target = 'class',train_size=0.9, data_split_shuffle = True, 
                 fold_strategy='stratifiedkfold', fold=10, 
                 silent = True, verbose = False, session_id = 123)


In [19]:
best_model = compare_models(sort='F1')

# catboost!

In [23]:
catboost = create_model('catboost')
tuned_catboost = tune_model(catboost)

In [33]:
final_catboost = finalize_model(tuned_catboost)

In [38]:
unseen_predictions = predict_model(final_catboost, data=df_unseen)
unseen_predictions.head()

liq_1      op_1      bs_1      fn_1  Unnamed: 0     Attr1     Attr3  \
0  0.010202 -0.044038 -0.012016 -0.983518        6513  0.052013  0.605560   
1 -0.057800 -0.043068 -0.007270  0.106401        6514  0.004142  0.004794   
2 -0.048837 -0.052328 -0.013823 -0.345820        6515  0.288930  0.713730   
3 -0.057284 -0.044175 -0.009534  1.534975        6516  0.156330 -0.392060   
4 -0.051659 -0.042144 -0.009554 -0.478918        6517  0.090630  0.324490   

      Attr5    Attr6     Attr7  ...   Attr58   Attr59  Attr60   Attr61  \
0  146.1800  0.18688  0.066273  ...  0.91819  0.00000  1.8966   5.2606   
1 -202.0200 -0.19273 -0.006737  ...  0.62859  0.45777  3.7113  14.7030   
2   14.5830  0.00000  0.360730  ...  0.85094  0.33994  3.7709   9.6891   
3 -101.3800 -0.72018  0.156330  ...  0.94250  0.00000  8.6625   5.4105   
4    1.8897 -0.34409  0.090630  ...  0.96172  0.00000  7.3060   9.6249   

    Attr62   Attr63   Attr64  class  Label   Score  
0   36.290  10.0580   2.0671    0.0    0.0  0.9991  
1  131.190   2.7822   1.6533    0.0    0.0  0.9766  
2   28.152  12.9650  23.1320    0.0    0.0  0.9834  
3  192.390   1.8972  53.3770    0.0    0.0  0.9857  
4   50.489   7.2293   6.2599    0.0    0.0  0.9979  

[5 rows x 62 columns]

In [39]:
from pycaret.utils import version, check_metric
check_metric(unseen_predictions['class'],unseen_predictions['Label'].astype('float32'), metric = 'F1')

0.6957

In [40]:
df_unseen['class'].value_counts()

0.0    326
1.0     15
Name: class, dtype: int64

# 정리

기본적으로 'catboost!' 전에는 동일, (즉 전처리, PCA 등이 동일, 기본 setting과 compared까지)

추가적으로 tuned하기 위해 catboost만 추가 진행하였다고 보면 됨. modeling_jcpark_v2에 catboost만 이어서 했다는 느낌.

결과는 위와 같음.

unseen_predictions는 f1 score 는 0.6957이 나옴